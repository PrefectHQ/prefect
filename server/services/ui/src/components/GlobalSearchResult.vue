<script>
/* eslint-disable vue/no-v-html */
// Added this line to disable to the v-html
// console warning since we know we can trust the
// data coming from the GraphQL server
export default {
  props: {
    searchResult: {
      required: true,
      type: Object,
      validator: result => {
        return (
          result.id &&
          result.name &&
          ['flow', 'flow_run', 'task', 'task_run'].includes(result.__typename)
        )
      }
    },
    // This is a reference to the parent passed in by the VAutoComplete
    // component. It contains important search highlighting methods
    // but we don't use it for much else. It's typed as an Object
    // and we're making sure the .genFilteredText() method exists on the
    // passed prop.
    parent: {
      type: Object,
      required: true,
      validator: parent => {
        return parent.genFilteredText
      }
    }
  },
  computed: {
    // Here we're just translating the
    // __typename reference to a human-readable format.
    // We could do this with a regex if we wanted.
    typeName() {
      switch (this.searchResult.__typename) {
        case 'flow':
          return 'Flow'
        case 'flow_run':
          return 'Flow Run'
        case 'task':
          return 'Task'
        case 'task_run':
          return 'Task Run'
        default:
          return ''
      }
    }
  }
}
</script>

<template>
  <div class="search-result">
    <v-list-item-content>
      <v-list-item-subtitle>{{ typeName }}</v-list-item-subtitle>
      <!-- the .genFilteredText() method calls a number of downstream methods
       on the parent component. This allows us to show where the user input
      matches the results we're returning to them.
      Since it returns HTML, we need to use the Vue HTML injector, otherwise
      it'll be inserted as plaintext. -->
      <v-list-item-title
        v-html="`${parent.genFilteredText(searchResult.name)}`"
      />
      <v-list-item-subtitle class="id-subtitle">
        {{ searchResult.id }}
      </v-list-item-subtitle>
    </v-list-item-content>
  </div>
</template>

<style lang="scss" scoped>
.search-result {
  cursor: pointer;
  min-width: 15rem;
}

.id-subtitle {
  font-size: 0.6rem !important;
}

.text--secondary {
  background-color: #123456 !important;
}
</style>
