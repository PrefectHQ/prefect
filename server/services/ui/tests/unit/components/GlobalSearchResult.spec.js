// Libraries
import Vuetify from 'vuetify'
import Vue from 'vue'

// Components
import GlobalSearchResult from '@/components/GlobalSearchResult'

// Utilities
import { createLocalVue, shallowMount } from '@vue/test-utils'

const localVue = createLocalVue()
Vue.use(Vuetify)

// Including this for now because the docs don't specify how
// we should be accessing passed-in parent props
// from vuetify components. Would need to combine
// these components to test the parent prop correctly, but
// we only need the specific method for now.
const genFilteredText = text => {
  return `<span>${text}</span>`
}

describe('GlobalSearch', () => {
  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const wrapperProps = () => {
    return {
      localVue,
      vuetify,
      propsData: {
        searchResult: {
          id: 'a45619e1 - 64ee - 49ac - a121 - b75fbfe5f375',
          name: 'Will I Retry ? Run me to find out.',
          __typename: 'flow'
        },
        parent: { genFilteredText: genFilteredText }
      }
    }
  }

  const wrapper = shallowMount(GlobalSearchResult, wrapperProps())

  it('renders a vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should contain the search-result class', () => {
    expect(wrapper.classes('search-result')).toBe(true)
  })

  it('should render the title element', () => {
    expect(wrapper.find('v-list-item-title-stub').exists()).toBe(true)
  })

  it('should render the subtitle element', () => {
    expect(wrapper.find('v-list-item-subtitle-stub').exists()).toBe(true)
  })
})
