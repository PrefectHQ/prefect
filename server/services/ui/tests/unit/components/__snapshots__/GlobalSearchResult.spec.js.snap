// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalSearch should match snapshot 1`] = `
<div
  class="search-result"
>
  <v-list-item-content-stub>
    <v-list-item-subtitle-stub>
      Flow
    </v-list-item-subtitle-stub>
     
    <v-list-item-title-stub />
     
    <v-list-item-subtitle-stub
      class="id-subtitle"
    >
      
      a45619e1 - 64ee - 49ac - a121 - b75fbfe5f375
    
    </v-list-item-subtitle-stub>
  </v-list-item-content-stub>
</div>
`;
